sim/dex-data.ts

```diff
@@ -90,105 +92,82 @@ export class BasicEffect {
 	sourceEffect: string;
 
-	constructor(data: AnyObject, moreData: AnyObject | null = null, moreData2: AnyObject | null = null) {
-		this.id = '';
-		this.name = '';
-		this.fullname = '';
-		this.effectType = 'Effect';
-		this.exists = true;
-		this.num = 0;
-		this.gen = 0;
-		this.isUnreleased = false;
-		this.shortDesc = '';
-		this.desc = '';
-		this.isNonstandard = false;
-		this.duration = this.duration;
-		// @ts-ignore
-		this.noCopy = !!this.noCopy;
-		// @ts-ignore
-		this.affectsFainted = !!this.affectsFainted;
-		this.status = this.status || undefined;
-		this.weather = this.weather || undefined;
-		this.drain = this.drain || undefined;
-		// @ts-ignore
-		this.flags = this.flags || {};
-		// @ts-ignore
-		this.sourceEffect = this.sourceEffect || '';
-
-		Object.assign(this, data);
-		if (moreData) Object.assign(this, moreData);
-		if (moreData2) Object.assign(this, moreData2);
-		this.name = Tools.getString(this.name).trim();
-		this.fullname = Tools.getString(this.fullname) || this.name;
-		if (!this.id) this.id = toId(this.name); // Hidden Power hack
-		// @ts-ignore
-		this.effectType = Tools.getString(this.effectType) || "Effect";
-		this.exists = !!(this.exists && this.id);
+	constructor(data: AnyObject) {
+		this.name = Tools.getString(data.name).trim();
+		this.id = data.id || toId(this.name); // Hidden Power hack
+		this.fullname = Tools.getString(data.fullname) || this.name;
+		this.effectType = Tools.getString(data.effectType) as EffectTypes || 'Effect';
+		this.exists = !!((data.hasOwnProperty('exists') ? data.exists : true) && this.id);
+		this.num = data.num || 0;
+		this.gen = data.gen || 0;
+		this.isUnreleased = data.isUnreleased || false;
+		this.shortDesc = data.shortDesc || '';
+		this.desc = data.desc || '';
+		this.isNonstandard = data.isNonStandard || false;
+		this.duration = data.duration;
+		this.noCopy = !!data.noCopy;
+		this.affectsFainted = !!data.affectsFainted;
+		this.status = data.status || undefined;
+		this.weather = data.weather || undefined;
+		this.drain = data.drain || undefined;
+		this.flags = data.flags || {};
+		this.sourceEffect = data.sourceEffect || '';
+		Object.assign(this, data); // TODO
 	}
+
 	toString() {
 		return this.name;
 	}
 }
 
+/** rule, source, limit, bans */
+type ComplexBan = [string, string, number, string[]];
+type ComplexTeamBan = ComplexBan;
+
 /**
  * A RuleTable keeps track of the rules that a format has. The key can be:
  * - '[ruleid]' the ID of a rule in effect
  * - '-[thing]' or '-[category]:[thing]' ban a thing
  * - '+[thing]' or '+[category]:[thing]' allow a thing (override a ban)
  * [category] is one of: item, move, ability, species, basespecies
  */
-export class RuleTable extends Map {
-	/** rule, source, limit, bans */
-	complexBans: [string, string, number, string[]][];
-	/** rule, source, limit, bans */
-	complexTeamBans: [string, string, number, string[]][];
+export class RuleTable extends Map { // TODO: For iteration? Can't we use composition?
```

Comment:
Why do we need to `extend Map` - usually [composition is prefered](https://en.wikipedia.org/wiki/Composition_over_inheritance)?

Reply:
Mostly because it was easier than reimplementing the Map interface.

Reply:
For the record I'm not opposed to composition, but this was just easier at the time.

Reply:
Gotcha